### PR TITLE
LFS: part 12

### PR DESCRIPTION
**Pull request by no92**

### CODE REVIEW

File: options/ansi/generic/stdlib-stubs.cpp [hunk 1]
```diff
@@ -297,14 +297,27 @@ void *bsearch(const void *key, const void *base, size_t count, size_t size,
 
 	return nullptr;
 }
+
+static int qsort_callback(const void *a, const void *b, void *arg) {
+	auto compare = (int (*)(const void *, const void *)) arg;
```
Comment (ikbenlike) [inline options/ansi/generic/stdlib-stubs.cpp:302]: would a `reinterpret_cast` here not be cleaner?

File: options/linux/generic/mntent-stubs.cpp [hunk 2]
```diff
@@ -1,16 +1,23 @@
...
 #include <mntent.h>
 #include <stdio.h>
+#include <limits.h>
 #include <string.h>
 #include <bits/ensure.h>
 
+static char *internal_buf;
+static size_t internal_bufsize;
+
+#define SENTINEL (char *)&internal_buf
```
Comment (ikbenlike) [inline options/linux/generic/mntent-stubs.cpp:12]: I am not a fan of these being global. is there no cleaner way to achieve the same or similar (for example passing in null to `getmntent_r` and checking against that instead)

Comment (no92) [reply options/linux/generic/mntent-stubs.cpp:16]: I have placed `internal_buf(size)` in an anonymous namespace.

File: options/glibc/include/sys/kd.h [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+#endif
+
+#include <linux/kd.h>
+
+#ifdef __undef_LINUX_TYPES_H
+# undef _LINUX_TYPES_H
+# undef __undef_LINUX_TYPES_H
+#endif
+
+#endif // _SYS_KD_H
```
Comment (Dennisbonke) [inline options/glibc/include/sys/kd.h:17]: We should use C style comments here.

File: options/posix/include/bits/posix/posix_stdio.h [hunk 4]
```diff
@@ -29,6 +30,21 @@ int vdprintf(int fd, const char *format, __gnuc_va_list args);
 
 char *fgetln(FILE *, size_t *);
 
+// GNU extensions
```
Comment (Dennisbonke) [inline options/posix/include/bits/posix/posix_stdio.h:33]: We should use C style comments here.

**Review by Dennisbonke — CHANGES_REQUESTED** [review]
LGTM outside 2 style comments.

File: options/ansi/generic/stdlib-stubs.cpp [hunk 5]
```diff
@@ -297,14 +297,27 @@ void *bsearch(const void *key, const void *base, size_t count, size_t size,
...
+
+	return compare(a, b);
+}
+
 void qsort(void *base, size_t count, size_t size,
 		int (*compare)(const void *, const void *)) {
+	return qsort_r(base, count, size, qsort_callback, (void *) compare);
+}
+
+void qsort_r(void *base, size_t count, size_t size,
```
Comment (Geertiebear) [inline options/ansi/generic/stdlib-stubs.cpp:312]: Could you add a test for qsort?

**Review by Geertiebear — CHANGES_REQUESTED** [review]
In general not a fan of the mtent code but we should probably have some mlibc wide code to do with successive file reading. So nothing to change about that for now. I would like to see a test for ``qsort()``, and it would also be nice if we could have one for the mtent functions (can be done by taking some dummy fstab file and verifying the code against it).

**Comment by ikbenlike** [comment]
I agree with Geert here. A test would be good, but some of the code quality changes are out of the scope of this PR. When that test gets added, LGTM as far as I'm concerned

**Review by Dennisbonke — APPROVED** [review]
LGTM

**Pull request merged**